### PR TITLE
Feat/AUT-345/predefined selectors for choices constraints

### PR DESCRIPTION
**Pull request by bziondik**
Related to: https://oat-sa.atlassian.net/browse/AUT-345
**Depends on:**
- [x] https://github.com/oat-sa/tao-item-runner-qti-fe/pull/254

image

**Acceptance criteria**
GIVEN a choiceInteraction with N defined number of choices:
AC1: Interaction properties
AC1a
GIVEN user browses item authoring tool
WHEN user inserts a choiceInteraction on the authoring canvas
THEN TAO shall present 2 radio-button groups with predefined selectors for [single / multiple choice] and constraints [None / Answer required / Other constraints] on Interaction Properties Panel.

The following constraints, when Single choice or Multiple choices is selected, MUST be transformed to attributes minChoices & maxChoices of the QTI choiceInteraction element

Single choice: 
   1. Constraint: None → minChoices = 0 / maxChoices = 1
   2. Constraint: Answer required → minChoices = 1 / maxChoices = 1
   3. Constraint: Other constraints → Radio-button is disabled

Multiple Choice: 
   4. Constraint: None → minChoices = 0 / maxChoices = 0
   5. Constraint: Answer required → minChoices = 1 / maxChoices = 0
   6. Constraint: Other constraints →
        a. minChoices = 1 / maxChoices = 2 (Default value)
        b. IF minChoices = 1 
             AND maxChoices = (N or Disabled) 
             THEN same as 5
        c. IF minChoices = 1 
             THEN maxChoices checkbox is disabled
              AND user is not able to uncheck and disable maxChoices constraint 
        d. IF minChoices = Disabled 
             AND maxChoices = N
             THEN same as #4
        e. IF maxChoices = Disabled 
             THEN minChoices  ≥ 2 (can never be = 1)

*minChoices from 1 to (N-1) (can never be = N)
  maxChoices from 2 to N (can never be = 1)
  maxChoices can never be < minChoices
  minChoices and maxChoices values can only be 0 when custom value widgets are Disabled, respectively
  minChoices and maxChoices values can never be Disabled at the same time

AC1b
GIVEN AC1a
WHEN user selects one of the 1,2,4 or 5 options as shown on AC1a
AND user saves the item
THEN TAO shall save the item with the corresponding values for minChoices & maxChoices

AC1c
GIVEN AC1a
WHEN user selects 6 option as shown on AC1a
THEN TAO shall display 2 spinner widgets for author to insert the custom values for minChoices & maxChoices, below the radio-button group
AND validate the inputs considering the rules on 6a,6b,6c and *

AC1d
GIVEN AC1a
WHEN user selects one of the 6a,6b, or 6c options as shown on AC1a
AND user saves the item
THEN TAO shall save the item with the corresponding values for minChoices & maxChoices

AC3: Special case when a single choice is defined
GIVEN AC1a
WHEN a choiceInteraction is configured with N = 1
THEN TAO shall disable Multiple choices option on the radio-button group

AC4: Tooltip messages
Constraints tooltip message: 

Define constraints on the number of choices required to form a valid response.
None: define no additional constraints.
Answer required: require the test taker to select at least one choice.
Other constraints: define custom minimum and/or maximum constraints (only available for multiple choices)

### CODE REVIEW

File: views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js [hunk 1]
```diff
@@ -101,46 +189,117 @@ define([
...
         //modify the checkbox/radio input appearances
-        widget.on('attributeChange', function(data){
-
-            var $checkboxIcons = widget.$container.find('.real-label > span');
+        widget.on('attributeChange', function (data) {
+            const $checkboxIcons = widget.$container.find('.real-label > span');
 
-            if(data.element.serial === interaction.serial && data.key === 'maxChoices'){
-                if(parseInt(data.value) === 1){
+            if (data.element.serial === interaction.serial && data.key === 'maxChoices') {
```
Comment (marpesia) [inline views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js:285]: There is an issue here. If I uncheck the `Max` constraint, it returns a value of `1` even is disabled. If I want to set a `Min` of 2 but with no `Max`, this is a problem. In this case, I have radio buttons too, but I want to check more than 1 option.
In the Mockup, this case should be a checkbox, not a radio button.
Can you please take a look?

**Review by marpesia — CHANGES_REQUESTED** [review]

File: views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js (hunk 1 above)
Comment (bziondik) [reply views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js:308]: @marpesia updated, please check

Comment (marpesia) [reply views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js:308]: It is working with the new code, but I found a new issue.
If you disabled the checkbox, it is still possible to change the Min/Max numbers and I've checked that it works.
image
And checking Figma I realized that we need to add a margin/padding on the left side
image

Comment (bziondik) [reply views/js/qtiCreator/widgets/interactions/choiceInteraction/states/Question.js:308]: _If you disabled the checkbox, it is still possible to change the Min/Max numbers and I've checked that it works._ - this is right behaviour
Styes are fixed

**Review by marpesia — APPROVED** [review]
LGTM

**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by Silvia-loza — APPROVED** [review]
All works as expected, Great job! :muscle:
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X]  Pull request's target is not ` master`
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

**Review by Silvia-loza — APPROVED** [review]
All looks good and works as expected, some many Ac's meet in this task great job!  :clap:
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X]  Pull request's target is not` master`
- [X] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful

**Review by marpesia — APPROVED** [review]
Good job!

**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Pull request merged**